File: packages/core/src/extract-accent.ts
/**
 * `extract-accent.ts` — resuelve color + posición del accent visual.
 *
 * Las recetas pueden pintar un borde lateral (accent) cuyo color se deriva
 * de un field con `behavior=color_hex` de la sección. La posición la decide
 * el publisher con una cascada de overrides; si todas son `auto`, se cae al
 * default que la receta provee (top en hero, left en compact_*).
 *
 * **Cascada de resolución de posición** (primer hit gana):
 *  1. `composition.accentPosition` (override global del publisher)
 *  2. `slot.appearance.accentPosition` del slot que contiene el field color
 *  3. `recipeDefault` (heurística cableada por la receta)
 *
 * Pure: sin side effects, sin acceso a DOM/CSS. Devuelve undefined si no
 * hay color hex válido en la sección — el caller renderea sin accent.
 *
 * Historia: extraído de `kromia-studio/src/components/album/recipes/
 * recipe-utils.tsx:223-261` en KRO-73 (migración B+).
 */

import type { SlotComposition, AccentSettings, FieldDefLike } from './types';

/**
 * Estructura mínima de la composition que esta función consume.
 * Compatible con `ViewComposition` (que tiene `slots` y opcionalmente
 * `accentPosition`).
 */
interface AccentSourceComposition {
  slots: Record<string, SlotComposition>;
  accentPosition?: 'top' | 'right' | 'bottom' | 'left' | 'none' | 'auto';
}

/**
 * Extrae color hex + posición del accent para una composition + item dados.
 *
 * @param composition   ViewComposition de la receta (puede ser undefined
 *                      si la sección no tiene composition declarada).
 * @param item          Fila de datos (key → value). Se consultan solo los
 *                      fields con behavior=color_hex.
 * @param fieldDefs     Definiciones de fields de la sección.
 * @param recipeDefault Posición por defecto que aplica la receta concreta
 *                      cuando todos los overrides son `auto`.
 * @returns             `{ color, position }` o `undefined` si no hay
 *                      color hex válido en `item`.
 */
export function extractAccentSettings(
  composition:   AccentSourceComposition | undefined,
  item:          Record<string, unknown>,
  fieldDefs:     FieldDefLike[],
  recipeDefault: 'top' | 'left',
): AccentSettings | undefined {
  // 1. Buscar color hex válido en cualquier field de la sección.
  let color: string | undefined;
  let colorFieldKey: string | undefined;
  for (const def of fieldDefs) {
    if (def.behavior !== 'color_hex') continue;
    const value = item[def.key];
    if (typeof value !== 'string') continue;
    const raw = value.trim();
    if (!/^#?[0-9a-f]{6}([0-9a-f]{2})?$/i.test(raw)) continue;
    color = raw.startsWith('#') ? raw : `#${raw}`;
    colorFieldKey = def.key;
    break;
  }
  if (!color || !colorFieldKey) return undefined;

  // 2. Resolver posición — orden de prioridad:
  //   a. composition.accentPosition (override global del publisher)
  //   b. slot.appearance.accentPosition (override per-slot legacy)
  //   c. recipeDefault (heurística por receta)
  let resolved: AccentSettings['position'] | 'auto' = composition?.accentPosition ?? 'auto';
  if (resolved === 'auto' && composition?.slots) {
    for (const slot of Object.values(composition.slots)) {
      if (!slot.fields.includes(colorFieldKey)) continue;
      const p = slot.appearance?.accentPosition;
      if (p) {
        resolved = p;
        break;
      }
    }
  }
  const position = resolved === 'auto' ? recipeDefault : resolved;
  // KRO-198 — `colorFieldKey` viaja en el resultado: el host suprime como celda el
  // slot que mapea ese campo cuando el acento está activo (no duplica swatch + raya).
  return { color, position, colorFieldKey };
}

File: packages/core_dart/lib/src/extract_accent.dart
/// `extract_accent.dart` — espejo Dart de `extract-accent.ts` (@kromia/core).
///
/// Resuelve color + posición del accent visual de una receta. El color sale de
/// un field con `behavior=color_hex` de la sección; la posición se resuelve con
/// la cascada `composition.accentPosition` > `slot.appearance.accentPosition` >
/// `recipeDefault`. Puro, sin side effects. Paridad 1:1 vigilada por el corpus
/// (`test/extract_accent_test.dart` espeja `tests/extract-accent.test.ts`).
library;

import 'composition.dart';
import 'field_def.dart';

/// Resultado de [extractAccentSettings] — color hex resuelto + posición.
/// Espejo de `AccentSettings` (types.ts).
class AccentSettings {
  final String color;

  /// 'top' | 'right' | 'bottom' | 'left' | 'none'.
  final String position;

  /// KRO-219 — ESTILO de la franja: 'bar'|'rounded'|'glow'|'gradient'|'ambient'.
  /// default 'bar'. Copiado de `composition.accentStyle` (NO altera color/posición).
  final String style;

  const AccentSettings({required this.color, required this.position, this.style = 'bar'});

  @override
  bool operator ==(Object other) =>
      other is AccentSettings &&
      other.color == color &&
      other.position == position &&
      other.style == style;

  @override
  int get hashCode => Object.hash(color, position, style);

  @override
  String toString() => 'AccentSettings(color: $color, position: $position, style: $style)';
}

/// Mismo patrón que `extract-accent.ts`: `#RRGGBB` o `#RRGGBBAA`, con o sin `#`.
final RegExp _hexRe =
    RegExp(r'^#?[0-9a-f]{6}([0-9a-f]{2})?$', caseSensitive: false);

/// Extrae color hex + posición del accent para una `composition` + `item`.
///
/// Devuelve `null` si no hay color hex válido en `item` (el caller renderiza
/// sin accent). [recipeDefault] es `'top'` o `'left'` según la receta.
AccentSettings? extractAccentSettings(
  ViewComposition? composition,
  Map<String, dynamic> item,
  List<FieldDefLike> fieldDefs,
  String recipeDefault,
) {
  // 1. Buscar color hex válido en cualquier field color_hex de la sección.
  String? color;
  String? colorFieldKey;
  for (final def in fieldDefs) {
    if (def.behavior != 'color_hex') continue;
    final value = item[def.key];
    if (value is! String) continue;
    final raw = value.trim();
    if (!_hexRe.hasMatch(raw)) continue;
    color = raw.startsWith('#') ? raw : '#$raw';
    colorFieldKey = def.key;
    break;
  }
  if (color == null || colorFieldKey == null) return null;

  // 2. Resolver posición — composition > slot override > recipeDefault.
  var resolved = composition?.accentPosition ?? 'auto';
  if (resolved == 'auto' && composition != null) {
    for (final slot in composition.slots.values) {
      if (!slot.fields.contains(colorFieldKey)) continue;
      final p = slot.appearance?.accentPosition;
      if (p != null) {
        resolved = p;
        break;
      }
    }
  }
  final position = resolved == 'auto' ? recipeDefault : resolved;
  // KRO-219 — el estilo se copia tal cual (default 'bar'); no afecta color/posición.
  return AccentSettings(color: color, position: position, style: composition?.accentStyle ?? 'bar');
}
